src/autopas/containers/linkedCells/HierarchicalGrid.h

```diff
@@ -0,0 +1,557 @@
+#include "autopas/containers/linkedCells/traversals/HGTraversalInterface.h"
+#include "autopas/iterators/ContainerIterator.h"
+
+namespace autopas {
+
+/**
+ * HierarchicalGrid class.
+ * This class stores multiple LinkedCells containers, each for different sizes of particles
+ * Traverses them all one by one, and cross hierarchy afterward
+ * @tparam Particle type of the Particle
+ */
+template <class Particle>
+class HierarchicalGrid : public ParticleContainerInterface<Particle> {
+ public:
+  /**
+   *  Type of the ParticleCell.
+   */
+  using ParticleCell = FullParticleCell<Particle>;
+
+  /**
+   *  Type of the Particle.
+   */
+  using ParticleType = Particle;
+
+  /**
+   * Constructor of the HierarchicalGrid class.
+   * @param boxMin
+   * @param boxMax
+   * @param cutoffs cutoffs for each level of the hierarchy
+   * @param skin Verlet skin
+   * @param rebuildFrequency the frequency the container will be definitely rebuilt
+   * @param cellSizeFactor cell size factor relative to cutoff
+   */
+  HierarchicalGrid(const std::array<double, 3> &boxMin, const std::array<double, 3> &boxMax,
+                   const std::vector<double> &cutoffs, const double skin, const unsigned int rebuildFrequency,
+                   const double cellSizeFactor = 1.0)
+      : ParticleContainerInterface<Particle>(skin),
+        _boxMin(boxMin),
+        _boxMax(boxMax),
+        _skin(skin),
+        _numLevels(cutoffs.size()),
+        _cellSizeFactor(cellSizeFactor),
+        _cacheOffset(DEFAULT_CACHE_LINE_SIZE / sizeof(size_t)),
+        _rebuildFrequency(rebuildFrequency),
+        _cutoffs(cutoffs) {
+    /*
+     * NOTE: In the future add auto cutoff? We need to know particles sizes beforehand.
+     * Not relevant for md simulations but can be useful for DEM simulations.
+     * Good heuristic for auto cutoffs:
+     * Let m = avg particle per cell, L = number of levels
+     * try to make m for each hierarchy level close, then minimize L*m
+     * prob works -> ternary search for m, for fixed m binary search over particle array sorted by size to find L
+     */
+    // resize iteration storage variables
+    _currentLevel.resize(autopas_get_max_threads() * _cacheOffset, 0);
+    _prevCells.resize(autopas_get_max_threads() * _cacheOffset, 0);
+
+    if (_cutoffs.empty()) {
+      utils::ExceptionHandler::exception("Error: Hierarchical Grid cutoffs vector is empty.");
+    }
+    // calculate max allowable cutoff
+    double maxLength = 1e15;
+    for (size_t i = 0; i < 3; ++i) {
+      maxLength = std::min(maxLength, boxMax[i] - boxMin[i]);
+    }
+    maxLength -= _skin;
+    // fix cutoffs if they are bigger than maxLength
+    for (auto &cutoff : _cutoffs) {
+      if (cutoff > maxLength) {
+        AutoPasLog(ERROR,
+                   "Cutoff {} is greater than the maximum length according to Box size - skin {}, lowering cutoff",
+                   cutoff, maxLength);
+        cutoff = maxLength;
+      }
+    }
```

Comment:
AutoPas shouldn't be used anyway with such a large cutoff compared to the box size. I would simply throw an exception - the simulation makes no sense and the cutoffs shouldn't be changed.